modules/javafx.graphics/src/main/java/com/sun/scenario/animation/LinearInterpolator.java

```diff
@@ -0,0 +1,211 @@
+            double xStep = (x1 - x0) / (j - i + 1);
+            double x = x0;
+            for (int k = i; k < j; ++k) {
+                x += xStep;
+                controlPoints[2 * k] = x;
+            }
+
+            i = j;
+        }
+    }
+
+    @Override
+    public double curve(double t) {
+        int n = controlPoints.length / 2;
+
+        // Only a single control point: always return its output value.
+        if (n == 1) {
+            return controlPoints[1];
+        }
+
+        // If t matches the x of one or more control points, return the y of the last such point.
+        double lastMatchY = Double.NaN;
+        for (int i = 0; i < n; ++i) {
+            double x = controlPoints[2 * i];
+            if (t < x) { // Xs are monotonically non-decreasing; no more matches possible
+                break;
+            }
+
+            if (t == x) {
+                lastMatchY = controlPoints[2 * i + 1];
+            }
+        }
+
+        if (!Double.isNaN(lastMatchY)) {
+            return lastMatchY;
+        }
+
+        double xFirst = controlPoints[0];
+        double yFirst = controlPoints[1];
+        double xLast  = controlPoints[2 * (n - 1)];
+        double yLast  = controlPoints[2 * (n - 1) + 1];
+        double ax, ay, bx, by;
+
+        // If t is smaller than any x, use the first segment.
+        if (t < xFirst) {
+            ax = xFirst;
+            ay = yFirst;
+            bx = controlPoints[2];
+            by = controlPoints[3];
+
+            if (ax == bx) {
+                return ay;
+            }
+        }
+        // If t is larger than any x, use the last segment.
+        else if (t > xLast) {
+            ax = controlPoints[2 * (n - 2)];
+            ay = controlPoints[2 * (n - 2) + 1];
+            bx = xLast;
+            by = yLast;
+
+            if (ax == bx) {
+                return by;
+            }
+        }
+        // t is between first.x and last.x (and not equal to any x).
+        else {
+            int indexA = -1; // last point with x < t
+            int indexB = -1; // first point with x > t
+
+            for (int i = 0; i < n; i++) {
+                double x = controlPoints[2 * i];
+                if (x < t) {
+                    indexA = i;
+                } else if (x > t) {
+                    indexB = i;
+                    break;
+                }
+            }
+
+            if (indexA < 0) {
+                ax = xFirst;
+                ay = yFirst;
+            } else {
+                ax = controlPoints[2 * indexA];
+                ay = controlPoints[2 * indexA + 1];
+            }
+
+            if (indexB < 0) {
+                bx = xLast;
+                by = yLast;
+            } else {
+                bx = controlPoints[2 * indexB];
+                by = controlPoints[2 * indexB + 1];
+            }
+        }
+
+        // Linearly interpolate (or extrapolate) along the segment (ax, ay) -> (bx, by).
+        if (ax == bx) {
```

Comment:
FP alert: is it possible for `ax != bx` yet the result of division on L203 to produce an infinity?

Reply:
I've added code to ensure that when an infinity is unavoidable, we at least get an infinity consistent with the line (and not, say, NaN). While this makes this interpolator locally consistent, we may still end up passing infinities into the animation system, so it's not a catch-all solution (that would be a completely different thing that we're not going to solve in interpolators).